File: leet_code/trees/postorder_traversal.rb
def postorder_traversal(root)
  return [] if !root
  result = []
  result.concat(postorder_traversal(root.left))
  result.concat(postorder_traversal(root.right))
  result << (root.val)
  result
end

# another solution

def postorder_traversal2(root, result = [])
    return result if !root
    postorder_traversal(root.left, result) if root.left
    postorder_traversal(root.right, result) if root.right
    result << root.val
end


# another solution

def postorder_traversal3(root)
  if !root
    return []
  else
    result = []
    traverse_rec = lambda do |node|
      traverse_rec.call(node.left) if node.left
      traverse_rec.call(node.right) if node.right
      result << node.val
    end
    traverse_rec.call(root)
    result
  end
end

File: leet_code/trees/postorder_traversal.js
const postorderTraversal = (root, result = []) => {
    if (!root) return result;
    if (root.left) postorderTraversal(root.left, result);
    if (root.right) postorderTraversal(root.right, result);
    result.push(root.val);
    return result;
};


// another solution

const postorderTraversal2 = root => {
  if (!root) return [];
  else {
    const result = [];
    const traverseRec = node => {
      if (node.left) traverseRec(node.left);
      if (node.right) traverseRec(node.right);
      result.push(node.val);
    };
    traverseRec(root);
    return result;
  }
};
